src/coreComponents/managers/Functions/TableFunction.cpp

```diff
@@ -177,129 +179,47 @@ void TableFunction::reInitializeFunction()
 }
 
+TableFunctionKernelWrapper TableFunction::createKernelWrapper() const
+{
+  return TableFunctionKernelWrapper( m_interpolationMethod,
+                                     m_coordinates.toViewConst(),
+                                     m_values.toViewConst(),
+                                     m_dimensions,
+                                     m_size.toViewConst(),
+                                     m_indexIncrement.toViewConst(),
+                                     m_corners,
+                                     m_numCorners );
+}
 
 real64 TableFunction::Evaluate( real64 const * const input ) const
 {
-  real64 result = 0.0;
-
-  // Linear interpolation
-  if( m_interpolationMethod == InterpolationType::Linear )
-  {
-    localIndex bounds[m_maxDimensions][2];
-    real64 weights[m_maxDimensions][2];
-
-    // Determine position, weights
-    for( localIndex ii=0; ii<m_dimensions; ++ii )
-    {
-      if( input[ii] <= m_coordinates[ii][0] )
-      {
-        // Coordinate is to the left of this axis
-        bounds[ii][0] = 0;
-        bounds[ii][1] = 0;
-        weights[ii][0] = 0;
-        weights[ii][1] = 1;
-      }
-      else if( input[ii] >= m_coordinates[ii][m_size[ii] - 1] )
-      {
-        // Coordinate is to the right of this axis
-        bounds[ii][0] = m_size[ii] - 1;
-        bounds[ii][1] = bounds[ii][0];
-        weights[ii][0] = 1;
-        weights[ii][1] = 0;
-      }
-      else
-      {
-        // Find the coordinate index
-        ///TODO make this fast
-        // Note: lower_bound uses a binary search...  If we assume coordinates are
-        // evenly spaced, we can speed things up considerably
-        auto lower = std::lower_bound( m_coordinates[ii].begin(), m_coordinates[ii].end(), input[ii] );
-        bounds[ii][1] = LvArray::integerConversion< localIndex >( std::distance( m_coordinates[ii].begin(), lower ));
-        bounds[ii][0] = bounds[ii][1] - 1;
-
-        real64 dx = m_coordinates[ii][bounds[ii][1]] - m_coordinates[ii][bounds[ii][0]];
-        weights[ii][0] = 1.0 - (input[ii] - m_coordinates[ii][bounds[ii][0]]) / dx;
-        weights[ii][1] = 1.0 - weights[ii][0];
-      }
-    }
-
-    // Calculate the result
-    for( localIndex ii=0; ii<m_numCorners; ++ii )
-    {
-      // Find array index
-      localIndex tableIndex = 0;
-      for( localIndex jj=0; jj<m_dimensions; ++jj )
-      {
-        tableIndex += bounds[jj][m_corners[jj][ii]] * m_indexIncrement[jj];
-      }
-
-      // Determine weighted value
-      real64 cornerValue = m_values[tableIndex];
-      for( localIndex jj=0; jj<m_dimensions; ++jj )
-      {
-        cornerValue *= weights[jj][m_corners[jj][ii]];
-      }
-      result += cornerValue;
-    }
-  }
-  // Nearest, Upper, Lower interpolation methods
-  else
+  // 1) copy the pointer input into an array
+  real64 scalarValue = 0;
+  stackArray1d< real64, maxDimensions > inputArray( m_dimensions );
```

Comment:
I think you could avoid copying the input values and just pass `input` directly to the `Compute` call below?

Reply:
Done, now I just pass `input` directly to the `compute` function.

On the same topic, but in the constitutive model, specifically here:
https://github.com/GEOSX/GEOSX/blob/4fcd1bd7eb78dd8ec9339330cded0caff1728a3d/src/coreComponents/constitutive/relativePermeability/TableRelativePermeability.hpp#L200-L203
is it legit to pass `&(dPhaseRelPerm_dPhaseVolFrac)[ip_water][ip_water]` as third argument to avoid the copy? I tried, it compiles and works on Quartz, but I am a little bit confused by what the `&&` sign does in the signature of the `TableFunctionKernelWrapper::compute` function.

Reply:
Yes, that should work too (and is a good idea).

The `&&` in that context (of a templated argument) is a "universal" reference, which simply means it can bind as an lvalue-ref or an rvalue-ref, depending on actual argument type that is deduced by compiler. That way you can pass via `derivatives` parameter either a local array (which is an lvalue) or some temporary (such as an ArraySlice generated in-flight from some indexing expression of a larger-dimensional ArrayView, which would be an r-value).

In the code you referenced, the type of `&(dPhaseRelPerm_dPhaseVolFrac)[ip_water][ip_water]` would be just `real64 *`, so it would get passed as `real64 * && derivatives` (rvalue-reference-to-pointer) which is perfectly legal.

Reply:
@klevzoff thanks for the explanation. I made the change in the code I referenced above, which now uses `&(dPhaseRelPerm_dPhaseVolFrac)[ip_water][ip_water]` to retrieve the derivatives from `TableFunction::KernelWrapper::compute`.